<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Импорт-библиотек-и-создание-соединения-с-базой-данных" data-toc-modified-id="Импорт-библиотек-и-создание-соединения-с-базой-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек и создание соединения с базой данных</a></span></li><li><span><a href="#Анализ-базы-данных" data-toc-modified-id="Анализ-базы-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ базы данных</a></span><ul class="toc-item"><li><span><a href="#Знакомство-с-базой-данных" data-toc-modified-id="Знакомство-с-базой-данных-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Знакомство с базой данных</a></span></li><li><span><a href="#Расчёт-количества-книг-вышедших-после-1-января-2000-года" data-toc-modified-id="Расчёт-количества-книг-вышедших-после-1-января-2000-года-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Расчёт количества книг вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Расчёт-количества-обзоров-и-средней-оценки" data-toc-modified-id="Расчёт-количества-обзоров-и-средней-оценки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Расчёт количества обзоров и средней оценки</a></span></li><li><span><a href="#Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(исключение-из-анализа-брошюр)" data-toc-modified-id="Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(исключение-из-анализа-брошюр)-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (исключение из анализа брошюр)</a></span></li><li><span><a href="#Определение-автора-с-самой-высокой-средней-оценкой-книг-(только-для-книг-с-50-и-более-оценками)" data-toc-modified-id="Определение-автора-с-самой-высокой-средней-оценкой-книг-(только-для-книг-с-50-и-более-оценками)-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Определение автора с самой высокой средней оценкой книг (только для книг с 50 и более оценками)</a></span></li><li><span><a href="#Расчёт-среднего-количества-обзоров-от-пользователей-(для-пользователей,-поставивших-более-50-оценок)" data-toc-modified-id="Расчёт-среднего-количества-обзоров-от-пользователей-(для-пользователей,-поставивших-более-50-оценок)-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Расчёт среднего количества обзоров от пользователей (для пользователей, поставивших более 50 оценок)</a></span></li></ul></li><li><span><a href="#Выводы-по-анализу" data-toc-modified-id="Выводы-по-анализу-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы по анализу</a></span></li></ul></div>

<font size = 5>**Финальный проект "SQL"**<font size>

# Введение

Цель проекта: изучив базу данных **с помощью запросов SQL** получить информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Во первом разделе содержится описание проекта.

Во втором разделе импортируются необходимые библиотеки и создаётся соединение с базой данных.

Анализ базы данных ведётся в разделе 3.

Выводы описаны в последнем разделе.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![title](u.png)

## Импорт библиотек и создание соединения с базой данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Анализ базы данных

### Знакомство с базой данных

In [18]:
tables = 'books', 'ratings', 'reviews', 'authors', 'publishers'
for table in tables:
    display(pd.io.sql.read_sql(f'SELECT * FROM {table} LIMIT 2', con = engine))

book_id  author_id                               title  num_pages  \
0        1        546                        'Salem's Lot        594   
1        2        465  1 000 Places to See Before You Die        992   

  publication_date  publisher_id  
0       2005-11-01            93  
1       2003-05-22           336

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs

publisher_id publisher
0             1       Ace
1             2  Ace Book

### Расчёт количества книг вышедших после 1 января 2000 года

In [8]:
f_query = '''SELECT COUNT(book_id) 
           FROM books 
           WHERE EXTRACT(YEAR FROM publication_date) >= 2000'''

In [9]:
pd.io.sql.read_sql(f_query, con = engine) 

count
0    821

### Расчёт количества обзоров и средней оценки

In [161]:
s_query = '''SELECT b.title, 
                    COUNT(r.review_id) count_reviews,
                    AVG(ra.rating) avg_rating
            FROM books b
            JOIN reviews r ON r.book_id = b.book_id
            JOIN ratings ra ON ra.book_id = b.book_id
            GROUP BY b.title
            ORDER BY avg_rating DESC'''

In [162]:
pd.io.sql.read_sql(s_query, con = engine)

title  count_reviews  \
0              School's Out—Forever (Maximum Ride  #2)             12   
1    The War of Art: Break Through the Blocks & Win...              4   
2            A Woman of Substance (Emma Harte Saga #1)              4   
3                    Misty of Chincoteague (Misty  #1)              4   
4    In the Hand of the Goddess (Song of the Liones...              6   
..                                                 ...            ...   
988  The World Is Flat: A Brief History of the Twen...             12   
989                                              Junky              4   
990                  His Excellency: George Washington              4   
991                                      Drowning Ruth              9   
992                               Harvesting the Heart              4   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
988        2.25  
989        2.00  
990        2.00  
991        2.00  
992        1.50  

[993 rows x 3 columns]

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (исключение из анализа брошюр)

In [46]:
t_query = '''SELECT p.publisher
            FROM publishers p
            JOIN books b ON p.publisher_id = b.publisher_id
            WHERE b.num_pages > 50
            GROUP BY p.publisher
            ORDER BY p.publisher DESC
            LIMIT 1
            '''

In [47]:
pd.io.sql.read_sql(t_query, con = engine)

publisher
0  Yearling Books

### Определение автора с самой высокой средней оценкой книг (только для книг с 50 и более оценками)

In [153]:
f_query = '''WITH
             util  AS    (         
                         SELECT a.author, b.title, nb.count, r.rating
                         FROM (SELECT book_id,
                                            COUNT(rating_id) 
                                    FROM ratings
                                    GROUP BY book_id) nb
                         JOIN books b ON b.book_id = nb.book_id
                         JOIN authors a ON b.author_id = a.author_id
                         JOIN ratings r ON b.book_id = r.book_id
                         WHERE nb.count > 50
                         )
             SELECT author
             FROM util
             GROUP BY author
             ORDER BY AVG(rating) DESC
             LIMIT 1
             '''

In [154]:
pd.io.sql.read_sql(f_query, con = engine)

author
0  J.K. Rowling/Mary GrandPré

### Расчёт среднего количества обзоров от пользователей (для пользователей, поставивших более 50 оценок)

In [155]:
t_query = '''WITH 
              needed AS (SELECT u.username, u.count
                         FROM (
                                SELECT username,
                                        COUNT(rating)
                                FROM ratings
                                GROUP BY username
                                ) u
                         WHERE u.count > 50)
                     
             SELECT AVG(count_reviews) avg_reviews
             FROM (
                     SELECT username,
                           COUNT(review_id) count_reviews
                     FROM reviews
                     WHERE username IN (SELECT username FROM needed)
                     GROUP BY username
                   ) review
                   
            '''

In [151]:
pd.io.sql.read_sql(t_query, con = engine)

avg_reviews
0    24.333333

## Выводы по анализу

1. За последнее два десятилетие в базу данных добавилось 821 книга.
2. Средняя оценка книга в базе данных колеблется от 1.5 до 5.
3. Самым продуктивным издательством является 'Yearling Books'.
4. Самым высокооцениваемым автором является J.K. Rowling/Mary GrandPré
5. Самые активные читатели в плане оценок книг сделали в среднем 24 обзора книг.